diff --git a/src/backend/utils/misc/tzparser.c b/src/backend/utils/misc/tzparser.c
index 6aaf7395ba8..45c517114e9 100644
--- a/src/backend/utils/misc/tzparser.c
+++ b/src/backend/utils/misc/tzparser.c
@@ -31,6 +31,7 @@
 #include "utils/memutils.h"
 #include "utils/tzparser.h"
 
+extern FILE *embedded_AllocateFile(const char *path, const char *mode);
 
 #define WHITESPACE " \t\n\r"
 
@@ -317,10 +318,8 @@ ParseTzFile(const char *filename, int depth,
 		return -1;
 	}
 
-	get_share_path(my_exec_path, share_path);
-	snprintf(file_path, sizeof(file_path), "%s/timezonesets/%s",
-			 share_path, filename);
-	tzFile = AllocateFile(file_path, "r");
+	strlcpy(file_path, filename, sizeof(file_path));
+	tzFile = embedded_AllocateFile(file_path, "r");
 	if (!tzFile)
 	{
 		/*
