diff --git a/src/backend/utils/fmgr/dfmgr.c b/src/backend/utils/fmgr/dfmgr.c
index 1366521f471..5fa4fced0a0 100644
--- a/src/backend/utils/fmgr/dfmgr.c
+++ b/src/backend/utils/fmgr/dfmgr.c
@@ -31,6 +31,13 @@
 /* signature for PostgreSQL-specific library init function */
 typedef void (*PG_init_t) (void);
 
+extern void *pg_load_external_function(const char *filename,
+									   const char *funcname,
+									   bool signalNotFound,
+									   void **filehandle);
+extern void *pg_lookup_external_function(void *filehandle,
+										 const char *funcname);
+
 /* hashtable entry for rendezvous variables */
 typedef struct
 {
@@ -75,8 +82,8 @@ static char *expand_dynamic_library_name(const char *name);
 static void check_restricted_library_name(const char *name);
 
 /* ABI values that module needs to match to be accepted */
-static const Pg_abi_values magic_data = PG_MODULE_ABI_DATA;
 
+const Pg_abi_values magic_data = PG_MODULE_ABI_DATA;
 
 /*
  * Load the specified dynamic-link library file, and look for a function
@@ -95,46 +102,7 @@ void *
 load_external_function(const char *filename, const char *funcname,
 					   bool signalNotFound, void **filehandle)
 {
-	char	   *fullname;
-	void	   *lib_handle;
-	void	   *retval;
-
-	/*
-	 * For extensions with hardcoded '$libdir/' library names, we strip the
-	 * prefix to allow the library search path to be used. This is done only
-	 * for simple names (e.g., "$libdir/foo"), not for nested paths (e.g.,
-	 * "$libdir/foo/bar").
-	 *
-	 * For nested paths, 'expand_dynamic_library_name' directly expands the
-	 * '$libdir' macro, so we leave them untouched.
-	 */
-	if (strncmp(filename, "$libdir/", 8) == 0)
-	{
-		if (first_dir_separator(filename + 8) == NULL)
-			filename += 8;
-	}
-
-	/* Expand the possibly-abbreviated filename to an exact path name */
-	fullname = expand_dynamic_library_name(filename);
-
-	/* Load the shared library, unless we already did */
-	lib_handle = internal_load_library(fullname);
-
-	/* Return handle if caller wants it */
-	if (filehandle)
-		*filehandle = lib_handle;
-
-	/* Look up the function within the library. */
-	retval = dlsym(lib_handle, funcname);
-
-	if (retval == NULL && signalNotFound)
-		ereport(ERROR,
-				(errcode(ERRCODE_UNDEFINED_FUNCTION),
-				 errmsg("could not find function \"%s\" in file \"%s\"",
-						funcname, fullname)));
-
-	pfree(fullname);
-	return retval;
+	return pg_load_external_function(filename, funcname, signalNotFound, filehandle);
 }
 
 /*
@@ -170,7 +138,7 @@ load_file(const char *filename, bool restricted)
 void *
 lookup_external_function(void *filehandle, const char *funcname)
 {
-	return dlsym(filehandle, funcname);
+	return pg_lookup_external_function(filehandle, funcname);
 }
 
 
